<root BTCPP_format="4">
    <BehaviorTree ID="MainTree">
        <Sequence name="SwapRoutine">
            
            <RetryUntilSuccessful num_attempts="5">
                <Action ID="GetCubesPoses" cube1_pose="{cube1_pose}" cube2_pose="{cube2_pose}" settling_time="3.0"/>
            </RetryUntilSuccessful>

            <Action ID="GoHome"/>

            <Action ID="GetCubeColor" cube_pose="{cube1_pose}" id="1" color="{cube1_color}"/>
            <Action ID="GetCubeColor" cube_pose="{cube2_pose}" id="10" color="{cube2_color}"/>

            <Fallback name="SwapWithRecovery">
                <Sequence name="SwapSteps">
                    <Action ID="MoveCube" from="{cube1_pose}" to="0.5; -0.1; 0.4; 1.0; 0.0; 0.0; 0.0" use_waypoint="false"/>
                    <Action ID="MoveCube" from="{cube2_pose}" to="{cube1_pose}" use_waypoint="false"/>
                    <Action ID="MoveCube" from="0.5; -0.1; 0.4; 1.0; 0.0; 0.0; 0.0" to="{cube2_pose}" use_waypoint="false"/>                    
                </Sequence>

                <Sequence name="RecoveryLogic">
                    <Action ID="GoHome"/>
                    <AlwaysFailure/>
                </Sequence>
            </Fallback>

            <Action ID="GoHome"/>

            <Action ID="GetCubeColor" cube_pose="{cube1_pose}" id="10" color="{cube2_color}" />

            <Action ID="GetCubesPoses" cube1_pose="{cube1_pose}" cube2_pose="{cube2_pose}" settling_time="3.0"/>

            <Action ID="Reset"/>


            <Action ID="DisplayCubeInfo" id="1" pose="{cube1_pose}" color="{cube1_color}"/>
            <Action ID="DisplayCubeInfo" id="10" pose="{cube2_pose}" color="{cube2_color}"/>
        </Sequence>
    </BehaviorTree>
</root>
